### PR TITLE
Add functions for figurate numbers & abundant, deficient, and perfect numbers

### PR DESCRIPTION
**Pull request by kellyi**
Hi!

I was working on writing a little library of number theory methods and came across this one, so I thought I'd just make a pull request to add a few of them.

This PR adds some functions to check for figurate numbers: triangle, square, pentagon, hexagon, heptagon, and octagon numbers. Each function returns a Boolean. I've also added functions to check for abundant, deficient, and perfect numbers -- also returning true/false.

Best,
Kelly

### CODE REVIEW

File: lib/is_hexagonal.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+/**
+  * Determines whether an integer is 'hexagonal'.
+  * @param {Number} an integer x to test
+  * @return {Boolean} Whether or not the x is hexagonal.
+  * @module number-theory
+  * @author Kelly Innes
+  */
+
+module.exports = function isHexagonal(n) {
+  return ((Math.sqrt((8 * n) + 1) + 1) / 4) % 1 === 0;
```
Comment (rsandor) [inline lib/is_hexagonal.js:16]: Perhaps a quick line note for the proof of this

File: lib/is_octagonal.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+/**
+  * Determines whether an integer is 'octagonal'.
+  * @param {Number} an integer x to test
+  * @return {Boolean} Whether or not the x is octagonal.
+  * @module number-theory
+  * @author Kelly Innes
+  */
+
+module.exports = function isOctagonal(n) {
+  return (((Math.sqrt((3 * n) + 1)) + 1) / 3) % 1 == 0;
```
Comment (rsandor) [inline lib/is_octagonal.js:16]: Same here, add a comment to where you got the formula (proof would be great). Ditto for the rest of these. Mostly I'd like to make reading the source educational :)

File: lib/is_perfect.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+/**
+  * Determines whether an integer is 'perfect'.
+  * @param {Number} an integer x to test
+  * @return {Boolean} Whether or not x is perfect.
+  * @module number-theory
+  * @author Kelly Innes
+  */
+
+module.exports = function isDeficient(n) {
+  var divisorsOfNumber = divisors(n);
```
Comment (rsandor) [inline lib/is_perfect.js:16]: Is there a faster algorithm known for this? I kinda doubt it but divisors is expensive :( worth checking out if you haven't already

**Comment by rsandor** [comment]
@kellyi - great work! These were definitely needed! Left a few comments, please ping me back with your findings or if you need further help. After that we will get it merged and versioned.

Thank you so much!

**Comment by rsandor** [comment]
@kellyi - Also, would you be so kind as to add unit tests to the `test/` directory for each of the new functions?

**Comment by kellyi** [comment]
Hi @rsandor -- 

Sure! I wrote tests for each of these new methods and included them in the `test/` directory. I also added a bit of explanation for how those algorithms worked: essentially they implement formula I found on the Wikipedia articles for each kind of figurate number, and they work by finding whatever figurate root of the given number then checking whether or not the result's a whole number. I linked the relevant Wikipedia articles for each, and they typically look like this:

https://en.wikipedia.org/wiki/Octagonal_number

(I implemented something like the "Test for Octagonal Number" for each of the figurates.)

I'm not sure whether there's a faster algorithm for the abundant/deficient/perfect number's thing -- I'd love to see it if so -- BUT the RosettaCode site looks to have a possibly quicker implementation of finding the divisors of an Int. I'm not sure whether that code's licensed under MIT or GPL or what, so I didn't directly include it, but it might be worth checking out regardless:

http://rosettacode.org/wiki/Factors_of_an_integer#JavaScript

Glad these look to be helpful! Thanks for creating the library in the first place! Let me know if you'd like me to make any other changes or updates or anything.

Best,
Kelly

**Comment by rsandor** [comment]
@kellyi - Great, I'll look at the changes when I am off work and hopefully get this merged tonight. Cheers :dancers:

**Comment by rsandor** [comment]
@kellyi - Merged and released in [v1.1.0](https://github.com/rsandor/number-theory/releases/tag/v1.1.0). Cheers!

**Comment by kellyi** [comment]
@rsandor Awesome. Thanks! :+1:

**Pull request merged**